presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/CachingHiveMetastore.java

```diff
@@ -738,95 +774,164 @@ public void dropRole(String role)
         }
         finally {
-            tablePrivilegesCache.invalidate(new UserTableKey(grantee, databaseName, tableName));
+            invalidateTablePrivilegesCache(grantee, databaseName, tableName);
         }
     }
 
     @Override
-    public Set<HivePrivilegeInfo> listTablePrivileges(String databaseName, String tableName, PrestoPrincipal principal)
+    public Set<HivePrivilegeInfo> listTablePrivileges(MetastoreContext metastoreContext, String databaseName, String tableName, PrestoPrincipal principal)
     {
-        return get(tablePrivilegesCache, new UserTableKey(principal, databaseName, tableName));
+        return get(tablePrivilegesCache, new KeyAndContext<>(withContext(metastoreContext), new UserTableKey(principal, databaseName, tableName)));
+    }
+
+    public Set<HivePrivilegeInfo> loadTablePrivileges(KeyAndContext<UserTableKey> loadTablePrivilegesKey)
+    {
+        return delegate.listTablePrivileges(loadTablePrivilegesKey.getContext(), loadTablePrivilegesKey.getKey().getDatabase(), loadTablePrivilegesKey.getKey().getTable(), loadTablePrivilegesKey.getKey().getPrincipal());
+    }
+
+    private static class KeyAndContext<T>
+    {
+        private final MetastoreContext context;
+        private final T key;
+        private final int hashCode;
+
+        public KeyAndContext(MetastoreContext context, T key)
+        {
+            this.context = requireNonNull(context, "identity is null");
+            this.key = requireNonNull(key, "key is null");
+            this.hashCode = Objects.hash(context, key);
+        }
+
+        public MetastoreContext getContext()
+        {
+            return context;
+        }
+
+        public T getKey()
+        {
+            return key;
+        }
+
+        @Override
+        public boolean equals(Object o)
+        {
+            if (this == o) {
+                return true;
+            }
+            if (o == null || getClass() != o.getClass()) {
+                return false;
+            }
+            KeyAndContext<?> other = (KeyAndContext<?>) o;
+            return hashCode == other.hashCode &&
+                    Objects.equals(context, other.context) &&
+                    Objects.equals(key, other.key);
+        }
+
+        @Override
+        public int hashCode()
+        {
+            return hashCode;
+        }
+
+        @Override
+        public String toString()
+        {
+            return toStringHelper(this)
+                    .add("context", context)
+                    .add("key", key)
+                    .toString();
+        }
     }
 
-    public Set<HivePrivilegeInfo> loadTablePrivileges(String databaseName, String tableName, PrestoPrincipal principal)
+    private MetastoreContext withContext(MetastoreContext context)
     {
-        return delegate.listTablePrivileges(databaseName, tableName, principal);
+        return isMetastoreImpersonationEnabled ? context : MetastoreContext.empty();
```

Comment:
I'm not sure if it is correct to drop the entire context. How about we drop only the username if the impersonation is disabled?